src/mds/cephfs_features.cc

```diff
@@ -23,6 +23,7 @@ static const std::array feature_names
   "multi_reconnect",
   "deleg_ino",
   "metric_collect",
+  "fscrypt",
```

Comment:
Ditto here. "alternate_dname" maybe? Note too that we're going to need for the MDS to track whether there is a crypto context associated with the inode, and maybe set a flag of some sort on the wire when there is and the client doesn't get an xattr blob in the trace. That feature might be better labeled as "fscrypt"

Reply:
Yeah, sure, will fix it.
BTW, does it make sense to lable as fscrypt version ? or juse one flag ?

Reply:
In you kclient patch [1], you were setting the "encryption.ctx", so we can check this in MDS and then set the 'inode_t::fscrypt = 1' for the root dir inode. And then for all the child CInodes will inherit from it.

[1] https://patchwork.kernel.org/patch/11774711/

Reply:
Every inode gets its own context. The context holds a nonce that's used to generate a unique per-inode key. Inheritance should be handled by the client. All we need here is for the MDS to notice that there is an encryption.ctx xattr associated with the inode and set an internal flag noting that it's encrypted, and somehow transmit that information in the inode info.

My main concern here is that the MDS could send inode info in a trace, or in readdir reply with no xattr information, even though it does have an encryption.ctx. At that point the client might not realize that an inode is encrypted when it should. If I'm wrong and xattrs are always transmitted then this is a non-issue. I couldn't tell from looking at the MDS code.

Later, we may also want the MDS to restrict the client from doing certain operations if it doesn't support fscrypt properly. I'm not sure what that should look like yet though, so I don't think we should try to implement it yet.

Reply:
> Every inode gets its own context. The context holds a nonce that's used to generate a unique per-inode key. 

Yeah, checked that code, it is.

> Inheritance should be handled by the client. All we need here is for the MDS to notice that there is an encryption.ctx xattr associated with the inode and set an internal flag noting that it's encrypted, and somehow transmit that information in the inode info in replies.

Okay

> 
> My main concern here is that the MDS could send inode info in a trace, or in readdir reply with no xattr information, even though it does have an encryption.ctx. At that point the client might not realize that an inode is encrypted when it should. If I'm wrong and xattrs are always transmitted then this is a non-issue. I couldn't tell from looking at the MDS code.
> 

It makes sense. 


> Later, we may also want the MDS to restrict the client from doing certain operations if it doesn't support fscrypt properly. I'm not sure what that should look like yet though, so I don't think we should try to implement it yet.